sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -126,6 +126,134 @@ case class MapKeys(child: Expression)
+    val arrayAT = children.map(_.dataType.asInstanceOf[ArrayType])
+    val fields = arrayAT.zipWithIndex.foldRight(List[StructField]()) { case ((arr, idx), list) =>
+      StructField(s"_$idx", arr.elementType, children(idx).nullable || arr.containsNull) :: list
+    }
+    StructType(fields)
+  }
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val genericArrayData = classOf[GenericArrayData].getName
+    val genericInternalRow = classOf[GenericInternalRow].getName
+
+    val evals = children.map(_.genCode(ctx))
+
+    val arrayTypes = children.map(_.dataType.asInstanceOf[ArrayType].elementType)
+
+    val arrVals = ctx.freshName("arrVals")
+    val arrCardinality = ctx.freshName("arrCardinality")
+    val biggestCardinality = ctx.freshName("biggestCardinality")
+
+    val inputs = evals.zipWithIndex.map { case (eval, index) =>
+      s"""
+        |${eval.code}
+        |if (!${eval.isNull}) {
+        |  $arrVals[$index] = ${eval.value};
+        |  $arrCardinality[$index] = ${eval.value}.numElements();
+        |} else {
+        |  $arrVals[$index] = null;
+        |  $arrCardinality[$index] = 0;
+        |}
+        |$biggestCardinality = Math.max($biggestCardinality, $arrCardinality[$index]);
+      """.stripMargin
+    }.mkString("\n")
+
+    val myobject = ctx.freshName("myobject")
+    val j = ctx.freshName("j")
+    val i = ctx.freshName("i")
+    val args = ctx.freshName("args")
+
+    val fillValue = evals.zipWithIndex.map { case (eval, index) =>
+      s"""
+      |      if ($j == ${index}) {
+      |        $myobject[$j] = ${CodeGenerator.getValue(s"$arrVals[$j]", arrayTypes(index), i)};
+      |      }
+      """.stripMargin
+    }
+
+    val fillValueSplitted = ctx.splitExpressions(
+      expressions = fillValue,
+      funcName = "fillValue",
+      arguments =
+        ("int", j) ::
+        ("Array[] Object", myobject) :: Nil)
+
+    ev.copy(s"""
+      |ArrayData[] $arrVals = new ArrayData[$numberOfArrays];
+      |int[] $arrCardinality = new int[$numberOfArrays];
+      |int $biggestCardinality = 0;
+      |$inputs
+      |Object[] $args = new Object[$biggestCardinality];
+      |for (int $i = 0; $i < $biggestCardinality; $i ++) {
+      |  Object[] $myobject = new Object[$numberOfArrays];
+      |  for (int $j = 0; $j < $numberOfArrays; $j ++) {
+      |    if ($arrVals[$j] != null && $arrCardinality[$j] > $i && !$arrVals[$j].isNullAt($i)) {
+      |      $fillValueSplitted
+      |    } else {
+      |      $myobject[$j] = null;
+      |    }
+      |  }
+      |  $args[$i] = new $genericInternalRow($myobject);
+      |}
+      |boolean ${ev.isNull} = false;
+      |$genericArrayData ${ev.value} = new $genericArrayData($args);
+    """.stripMargin)
+  }
+
+  override def eval(input: InternalRow): Any = {
+    val inputArrays = children.map(_.eval(input).asInstanceOf[ArrayData])
+    val arrayTypes = children.map(_.dataType.asInstanceOf[ArrayType].elementType)
```

Comment:
can you move this as a private member in order to avoid re-evaluation at every row?